src/native/http_request_response.c

```diff
@@ -0,0 +1,481 @@
+        callback->java_crt_http_callback_handler,
+        s_crt_http_stream_handler.onResponseHeadersDone,
+        callback->java_http_stream,
+        jHasBody);
+
+    if ((*env)->ExceptionCheck(env)) {
+        // Close the Connection if the Java Callback throws an Exception
+        aws_http_connection_close(aws_http_stream_get_connection(stream));
+        return;
+    }
+}
+
+static void s_on_incoming_body_fn(
+    struct aws_http_stream *stream,
+    const struct aws_byte_cursor *data,
+    /* NOLINTNEXTLINE(readability-non-const-parameter) */
+    size_t *out_window_update_size,
+    void *user_data) {
+
+    struct http_stream_callback_data *callback = (struct http_stream_callback_data *)user_data;
+    JNIEnv *env = aws_jni_get_thread_env(callback->connection->jvm);
+
+    jobject jByteBuffer = aws_jni_byte_buffer_copy_from_cursor(env, data);
+    jint window_increment = (*env)->CallIntMethod(
+        env,
+        callback->java_crt_http_callback_handler,
+        s_crt_http_stream_handler.onResponseBody,
+        callback->java_http_stream,
+        jByteBuffer);
+
+    // Verify Window Increment is within bounds
+    assert(0 <= window_increment);
+    assert(window_increment <= *out_window_update_size);
+
+    if ((*env)->ExceptionCheck(env)) {
+        // Close the Connection if the Java Callback throws an Exception
+        aws_http_connection_close(aws_http_stream_get_connection(stream));
+        return;
+    }
+
+    // We can check the ByteBuffer read position to verify that the userThread actually read all the data they claimed
+    // to be able to read.
+    int amt_read = aws_jni_byte_buffer_get_position(env, jByteBuffer);
+    (void)amt_read;
+    assert(amt_read == data->len);
+
+    *out_window_update_size = window_increment;
+}
+
+static void s_on_stream_complete_fn(struct aws_http_stream *stream, int error_code, void *user_data) {
+
+    struct http_stream_callback_data *callback = (struct http_stream_callback_data *)user_data;
+    JNIEnv *env = aws_jni_get_thread_env(callback->connection->jvm);
+
+    jint jErrorCode = error_code;
+    (*env)->CallVoidMethod(
+        env,
+        callback->java_crt_http_callback_handler,
+        s_crt_http_stream_handler.onResponseComplete,
+        callback->java_http_stream,
+        jErrorCode);
+
+    if ((*env)->ExceptionCheck(env)) {
+        // Close the Connection if the Java Callback throws an Exception
+        aws_http_connection_close(aws_http_stream_get_connection(stream));
+        return;
+    }
+
+    http_stream_callback_release(env, callback);
+}
+
+enum aws_http_outgoing_body_state s_stream_outgoing_body_fn(
+    struct aws_http_stream *stream,
+    struct aws_byte_buf *dst,
+    void *user_data) {
+
+    struct http_stream_callback_data *callback = (struct http_stream_callback_data *)user_data;
+    JNIEnv *env = aws_jni_get_thread_env(callback->connection->jvm);
+
+    jobject jDirectByteBuffer = aws_jni_direct_byte_buffer_from_byte_buf(env, dst);
```

Comment:
See above about NewDirectByteBuffer, this is going to be a problem if someone keeps a reference to this ByteBuffer, this should probably be a byte[] copy.

Reply:
Ok, after discussion in chat, forget this, we're going to leave it up to the user.